### DISCUSSION TITLE
Alarms

### DISCUSSION THREAD

**Discussion by aregcoretigo** [ottowayi/pycomm3#204] [Q&A]
Hi,

First of all thank you for looking at my questions and for your continuous support. 
I am looking for a way to read alarm values from Control and Compact Logix Rockwell PLC (those supporting it should have firmware version 31 or newer). I couldn't find any information on that, is there a way to list tag alarms and read their values using pycomm3?

Thanks very much for your help,

Alex

**Comment by ottowayi** [comment]
If they are tags in the controller, then it should work.  Are you referring to the ALARM_ANALOG/DIGITAL tag types in the PLC?  I also think you can't use the full structure reading/writing with them because some attributes have the external access set to None.  So you may have to handle each attribute individually.  You can filter the tag list pretty, for example:

`alarm_tags = [tag for tag, info for in plc.tags.items() if info['data_type_name'].startswith('ALARM')]`

**Reply by Xiadais** [reply]
I believe @aregcoretigo is referring to Logix 5000 Tag-Based Alarming, which is a fairly new feature to Logix controllers - only 5380s, 5480s, and 5580s, and only since controller firmware v31. They use an @ syntax. For example, `Tag1.@Alarms.Nameofthealarm.Condition`.

I've used them a little bit, but I don't really know how they work from a technical perspective. I think it's just extended tag properties that get automatically, periodically evaluated by the controller. I guess the only part that really matters is that they're _probably_ read/written as extended tag properties. It is definitely possible to read/write them over a network - they can be used in PanelView 5000s or FactoryTalk View SE, and they're a key part of their new PlantPAx 5.0 system with "P" controllers and "process" instructions baked into controller firmware, communicating with View SE through FactoryTalk Linx.

FactoryTalk Linx 6.0+ is required to read/write the "@Alarm" alarms from Logix controllers, so it might be tricky to track down how the packets work.

**Reply by ottowayi** [reply]
Ohh, I understand... umm they probably don't work because it's a bullshit proprietary feature.  I remember a long time ago trying to reverse engineer the extended properties to read the tag description, but I didn't get it figured out.  I recall discovering it was using a _write_ service to read those properties and it was nothing like reading the tag itself.   If anyone has the time and patience to get it figured out I would gladly add support, but I'm not able to right now (I don't have a new enough controller to test against) and Rockwell is actively hostile towards anything not in their overpriced walled garden.
